FWCore/Framework/src/StreamSchedule.cc

```diff
@@ -1189,7 +1189,12 @@ namespace edm {
         ServiceRegistry::Operate operate(weakToken.lock());
 
         if (iPtr) {
-          //this is used to prioritize this error over one
+          // free previous value of pathErrorPtr, if any
+          auto currentPtr = pathErrorPtr->load();
+          if (currentPtr) {
+            delete currentPtr;
+          }
+          // this is used to prioritize this error over one
           // that happens in EndPath or Accumulate
           pathErrorPtr->store(new std::exception_ptr(*iPtr));
```

Comment:
I don't believe pathErrorPtr can be access concurrently. It is either set here in `pathsDone` (I don't think it can be non-null actually) or it is set later in `allPathsDone`. Where `allPathsDone` is guaranteed to be called only after `pathsDone` has finished since the lambda passed to `pathsDone` is holding `allPathsHolder`.

If that analysis is correct, there isn't any need for the use of any `exchange`.

Reply:
Could `pathErrorHolder` be made non-`std::atomic` then too?

Reply:
The `pathErrorPtr` could be set in one thread and read in another so needs synchronization. The `allPathsHolder` goes out of scope before the task running `pathsDone` does its final synchronization. Therefore `allPathsHolder` might run before that final synchronization happens. The use of `std::atomic` guarantees that the pointer is properly synchronized even in that case.

Reply:
Ok, so even though the `pathErrorPtr` is accessed by one thread at a time (that permits the "sloppy" use of `atomic` API), the `atomic` operations serve as a memory fence that guarantee the value to be the same in `pathsDone` and `allPathsDone` even if they would be run by different threads. Would this be phrasing be correct?

Reply:
Notice that even with this change analyzer still reports "Potential leak of memory pointed to by `pathErrorPtr`"...

Reply:
> Notice that even with this change analyzer still reports "Potential leak of memory pointed to by `pathErrorPtr`"...

I can believe that. The static analyzer doesn't know the context @Dr15Jones explained above. I would expect e.g. an `exchange()` and asserting the return value (i.e. the old value) to be `nullptr` to silence the warning.

Reply:
No, analyzer still complains.